src/library_async.js

```diff
@@ -309,16 +312,19 @@ mergeInto(LibraryManager.library, {
           }
           assert(!EmterpreterAsync.postAsync);
           EmterpreterAsync.postAsync = post || null;
-          Module['emterpret'](stack[0]); // pc of the first function, from which we can reconstruct the rest, is at position 0 on the stack
+          var ret;
+          if (EmterpreterAsync.restartFunc) ret = EmterpreterAsync.restartFunc();
```

Comment:
I did move the `restartFunc` clearing into the `resume` function in https://github.com/kripken/emscripten/pull/6586/commits/04f17c4802b7a1e3d051d1ec01fa15df7b48fd9c#diff-e1dcd8e7fa0139253f677464734f4f4cR326
(Note that `restartFunc` can be called multiple times, when there are multiple async operations).

Reply:
Sorry, not sure I understand what you mean by multiple async operations? I don't think we support multiple async operations in flight at the same time. (And if we did, I think we'd need a stack, not just a single restartFunc?)

Reply:
I mean multiple sequential async operations. If the code starts another async operation during the restart, we can't clear `restartFunc` yet (just like the finalizers are only called when `EmterpreterAsync.state === 0`).

Reply:
I see. How about saving the restartFunc to a local, nulling the restartFunc, and calling that local?

Reply:
I don't think that works, when `handle` is called the next time the local won't be available.

Reply:
Then I may still be misunderstanding something. What I meant is this:
```
if (EmterpreterAsync.restartFunc) {
  // Calling the restartFunc may start another async operation, which
  // assigns to the restartFunc, so clear it immediately in preparation.
  var temp = EmterpreterAsync.restartFunc;
  EmterpreterAsync.restartFunc = null;
  temp();
}
```

Reply:
(So I don't follow why the local would be needed the time time?)

Reply:
In either case, we should add a test for this specific case of starting another async operation immediately at the end of one - this is a tricky enough thing that we should have test coverage for it.

Reply:
But would those `emscripten_sleep` calls will not create a restartFunc? I thought only the new async ccall does?

It seems like the risk is mixing an async ccall with an `emscripten_sleep`, but that seems like something we should assert against anyhow?

Reply:
As I said, `restartFunc` is only set when going through `ccall`, and `emscripten_sleep` does not call `ccall` (it calls `EmterpreterAsync.handle`). And we already have those asserts (The one [here](https://github.com/jscissr/emscripten/blob/55783d98ca68caa540509d8ac372103543ef24a4/src/library_async.js#L355) and the additional one I added [here](https://github.com/kripken/emscripten/pull/6586/commits/c90ae46ad786bbedbc1ba70e13f8713ecb5bb04f#diff-5df1ca8b00952a4230cf0be8d8bf3682R178).

Reply:
Sorry, I'm confused: a few comments above, you wrote this:
```
If you have for example this C code:

emscripten_sleep(100);
emscripten_sleep(100);

After the first timeout, restartFunc is called, but because there is another async operation, we can't clear restartFunc yet.
```
That comment seems to connect `emscripten_sleep` and `restartFunc` - did I misunderstand you?

That comment was in response to my suggestion here https://github.com/kripken/emscripten/pull/6586#discussion_r197942002 which I still think is a safer, cleaner way to reset the restartFunc - but could be I'm missing something.

Reply:
The suggestion in https://github.com/kripken/emscripten/pull/6586#discussion_r197942002 doesn't work, if you don't believe me you can run the test (`asmi.test_async`).
The reason is that during an async ccall, `restartFunc` is set once, is called once or multiple times and then it is cleared, when the async call returns. If you already clear `restartFunc` after it is called the first time, that doesn't work.

Reply:
The `emscripten_sleep` was an example where it fails with your patch.
When `emscripten_sleep` is called from C code, `emscripten_sleep` will call `EmterpreterAsync.handle`. when the timeout expires, `resume` is called, and `resume` calls `restartFunc`.

Reply:
It has nothing to do with the test calling `ccall` twice, the problem is that the second `ccall` (to `floatf`) doesn't return a value, which is because `restartFunc` is cleared too early with your patch.

> Second, why is the restartFunc set more than once? Do we set it for every function on the stack during the async operation? That seems unnecessarily complex, but I may be missing something here?

It's not *set* more than once during an asnyc ccall, but it may be *called* multiple times.

Reply:
I think I see, thanks. So the issue is doing an async `emscripten_sleep` during an async `ccall`.

In general, nesting one async operation inside another worries me. I think this PR is the first time that we allowed that?

If we do want to allow such nesting, it would be more intuitive - and I hope less error-prone - to maintain a stack of async operations in flight, including for each their restartFuncs and asyncFinalizers. That could support all possible nestings, instead of just allowing one kind (`emscripten_sleep` inside `ccall`, but not `ccall` inside `ccall`).

However, it would be simpler if we didn't support nesting async operations at all - how important is that use case?

Reply:
Sorry, but it looks like you are completely misunderstanding something here. Async ccalls are only async because they do an async operation, and `emscripten_sleep` is one of them. This was already possible before this PR, all this PR adds is support for return values in async ccalls.

Reply:
Yeah, sorry - I had some confusion, and wrote something very unclear. What I meant is: this PR makes it possible to do an async `emscripten_sleep` inside the new async `ccall` functionality added here. But, I believe we disallow an async `ccall` from being called during an `emscripten_sleep` or another async `ccall`. There's an asymmetry there that worries me.

But, maybe the right way to look at this is that the asymmetry is intentional?

Reply:
> the new async ccall functionality added here.

What are you referring to here? I'm not quite sure I understand you correctly.

> If we do want to allow such nesting, it would be more intuitive - and I hope less error-prone - to maintain a stack of async operations in flight, including for each their restartFuncs and asyncFinalizers. That could support all possible nestings, instead of just allowing one kind (emscripten_sleep inside ccall, but not ccall inside ccall).

Yes, something like this would be possible. But it's not so simple to do, and also an independent feature from what this PR was intended for (support for return values).

Reply:
I meant the new functionality added in this entire PR, and how it relates to other async operations.

Overall, I think I have a good-enough understanding of this now, thanks again for your work and patience! I think we can merge this in. I'll do another pass at docs and comments myself later on as a followup.